Stop level obs details script

bsaul opened the pull request.

bsaul commented:
@IJBG - I'm working through extracting stop-level information from the observation details. I have a question regarding strings that don't split to exactly 20 stops, in particular using `pad_or_truncate`. As I understand, for the 31 cases where the string splits into 21 stops, `pad_or_truncate` simply truncates at 20. There's at least one case, though, where this cuts out data. Do you think it would be feasible to correct the data? I don't really see how without checking with the observer, but just trying to brainstorm here...

bsaul commented on R/stop-level_obs_comments.R:
```diff
@@ -1,121 +1,130 @@
+
+
+  
+  # # find rows where the output does not match count.
```
Add back error catching/handling. I've been thinking that we want to add a logging functionaliity to our build process because there are multiple places in the build where we want to identify possible problems.

IJBG commented on R/stop-level_obs_comments.R:
```diff
@@ -1,121 +1,132 @@
-    "To follow up see:",
-    "inst/species_comments_errors.csv"
-  ))
+process_obs_details <- function(ebird) {
```
Wait, woah - does this replace all three of the different sp_com handling functions?

bsaul replied:
Yes.

IJBG commented:
> @IJBG - I'm working through extracting stop-level information from the observation details. I have a question regarding strings that don't split to exactly 20 stops, in particular using `pad_or_truncate`. As I understand, for the 31 cases where the string splits into 21 stops, `pad_or_truncate` simply truncates at 20. There's at least one case, though, where this cuts out data. Do you think it would be feasible to correct the data? I don't really see how without checking with the observer, but just trying to brainstorm here...

Doesn't it cut out an observation of 1 bird? Still important, but I think the feasible correction would be trimming the species_comment in ebird and correcting it there instead. Not sure how we would do it otherwise without checking with the observer. Do you know what year it's from? Alternative correction is changing that last stop to +1 bird (again within ebird I think)

bsaul commented:
> > @IJBG - I'm working through extracting stop-level information from the observation details. I have a question regarding strings that don't split to exactly 20 stops, in particular using `pad_or_truncate`. As I understand, for the 31 cases where the string splits into 21 stops, `pad_or_truncate` simply truncates at 20. There's at least one case, though, where this cuts out data. Do you think it would be feasible to correct the data? I don't really see how without checking with the observer, but just trying to brainstorm here...
> 
> Doesn't it cut out an observation of 1 bird? Still important, but I think the feasible correction would be trimming the species_comment in ebird and correcting it there instead. Not sure how we would do it otherwise without checking with the observer. Do you know what year it's from? Alternative correction is changing that last stop to +1 bird (again within ebird I think)

Yes, just one. I agree that the preferred approach is to fix in ebird.

bsaul commented:
@IJBG - I added logging for parse errors/warnings. For now, these go to the default logging destination (the console).

Output looks like:

```
> dx <- process_obs_details(dt)
WARN [2024-10-07 16:41:58] S37625135 American Crow: observation details parsed to length == 19.
WARN [2024-10-07 16:41:58] S37534504 Blue Grosbeak: observation details parsed to length == 19.
ERROR [2024-10-07 16:41:58] S46457285 Eastern Wood-Pewee: observation details parsed to length > 20.
ERROR [2024-10-07 16:41:59] S29843524 Mourning Dove: observation details parsed to length > 20.
ERROR [2024-10-07 16:41:59] S29843524 Chimney Swift: observation details parsed to length > 20.
ERROR [2024-10-07 16:41:59] S29843524 Great Blue Heron: observation details parsed to length > 20.
ERROR [2024-10-07 16:41:59] S29843524 Red-bellied Woodpecker: observation details parsed to length > 20.
WARN [2024-10-07 16:41:59] S57003189 Red-bellied Woodpecker: observation details parsed to length == 19.
```

Is this sufficient for now?

bsaul commented:
> @IJBG - I added logging for parse errors/warnings. For now, these go to the default logging destination (the console).
> 
> Output looks like:
> 
> ```
> > dx <- process_obs_details(dt)
> WARN [2024-10-07 16:41:58] S37625135 American Crow: observation details parsed to length == 19.
> WARN [2024-10-07 16:41:58] S37534504 Blue Grosbeak: observation details parsed to length == 19.
> ERROR [2024-10-07 16:41:58] S46457285 Eastern Wood-Pewee: observation details parsed to length > 20.
> ERROR [2024-10-07 16:41:59] S29843524 Mourning Dove: observation details parsed to length > 20.
> ERROR [2024-10-07 16:41:59] S29843524 Chimney Swift: observation details parsed to length > 20.
> ERROR [2024-10-07 16:41:59] S29843524 Great Blue Heron: observation details parsed to length > 20.
> ERROR [2024-10-07 16:41:59] S29843524 Red-bellied Woodpecker: observation details parsed to length > 20.
> WARN [2024-10-07 16:41:59] S57003189 Red-bellied Woodpecker: observation details parsed to length == 19.
> ```
> 
> Is this sufficient for now?

I just added additional trace information:

```r
logger::log_threshold(logger::TRACE)
# dt <- create_stop_level()
# dt
dt <- get_ebird_data()
dx <- process_obs_details(dt)
dx
```
yields (e.g:

```r
ERROR [2024-10-08 10:42:17] S29843524 Northern Cardinal: observation details parsed to length > 20.
TRACE [2024-10-08 10:42:17] S29843524 Northern Cardinal: counts truncated at first 20 stops
ERROR [2024-10-08 10:42:17] S29843524 Indigo Bunting: observation details parsed to length > 20.
TRACE [2024-10-08 10:42:17] S29843524 Indigo Bunting: counts truncated at first 20 stops
ERROR [2024-10-08 10:42:17] S46295712 Indigo Bunting: observation details parsed to length > 20.
TRACE [2024-10-08 10:42:17] S46295712 Indigo Bunting: counts truncated at first 20 stops
``

bsaul commented:
@IJBG - I'm going to go ahead and merge this to the 93 branch in a bit -- I need this work in order to really get going on #123.

IJBG commented:
You can go ahead and merge, this all looks good! That error logging is perfect
